be able to sort dashboard articles (#1837)

omenking opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
```
Given I am a logged in user
And I am on my dashboard
And I am viewing articles
When I sort by articles by X
Then it should sort by X
```
## Related Tickets & Documents

In the original ticket I have provided a detailed journey to how I implemented this feature:
https://github.com/thepracticaldev/dev.to/issues/1837

In the `Dashboards#show`, there is both articles and followers code. I wonder if the follower code is still being utilized because there are two actions more appropriately named below the `#show` action.

I would recommend this `#show` action to be refactored to be single-focused in functionality and additional actions and routes spun up because from testing perceptive too much is going on here and the limits imposed by the linter does not allow for an elegant refactor.


## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![As they fly by, fly so high](https://66.media.tumblr.com/43bf1ab470e575c8bf694c28b12cc662/tumblr_ne0x88pubU1tooympo1_400.gif)

omenking commented:
So its CodeClimate complains

<img width="741" alt="Screen Shot 2019-04-08 at 12 57 30 PM" src="https://user-images.githubusercontent.com/7776/55742546-f3990300-59fd-11e9-9651-a6e68ffed069.png">

So here you can see how I ensure injection won't happen:
```
    unless %w[user_followers organization_user_followers].include?(params[:which])
      sorts = %w[created_at comments_count positive_reactions_count page_views_count]
      params[:sort] ||= "created_at"
      sort_by = "created_at"
      sort_by = params[:sort] if params[:sort] && sorts.include?(params[:sort])
    end
````

So to make it clear, only whitelisting what is allowed.

```
sorts = %w[created_at comments_count positive_reactions_count page_views_count]
sorts.include?(params[:sort])
```

To get Code Climate to not complain I could do a `case` or `if` or I could see if it would not complain if I create a `scope`

please advise what you would like.

rhymes commented on app/controllers/dashboards_controller.rb:
```diff
@@ -5,16 +5,24 @@ class DashboardsController < ApplicationController
     fetch_and_authorize_user
+
+    unless %w[user_followers organization_user_followers].include?(params[:which])
+      sorts = %w[reated_at comments_count positive_reactions_count page_views_count]
```
A `c` is missing

rhymes commented:
Since this is a user facing feature would you kindly add some tests? At least to check what happens if you give "sort by" something unknown vs something known and to make sure that the injection cannot happen.

Another possible improvement is to separate the name of the fields in the API from the name of the fields in the DB.

So instead of `sort=created_at` you could implement the following:

* `sort=creation-desc` and `sort=creation-asc` that maps to `created_at`
* `sort=views-desc` and `sort=views-asc` that maps to `page_views_count`
* `sort=reactions-desc` and `sort=reactions-asc` that maps to `reactions_count`
* `sort=comments-desc` and `sort=comments-asc` that maps to `reactions_count`

this way the names of the fields are not tied to the URL and you get the sort direction for free in case in the future you want to add the order's direction

omenking commented:
## Test Suite Dilemma 
Investing the codebase I do not see a place to spec controller actions as I would expect this to be in `spec/controllers` directory. 

<img width="334" alt="Screen Shot 2019-04-08 at 4 52 53 PM" src="https://user-images.githubusercontent.com/7776/55756196-ea6c5e00-5a1e-11e9-8c78-dcb3fa1e0910.png">

I also did not see any `features` directory so I could write user acceptance tests. I double checked the app directory to see many if user acceptance test were written elsewhere perhaps with javascript but I could not determine such a place.

## Suggested Changes

Since injection is of a cocern I will write a more verbose solution.
I can as well as you suggest obscurify the attribute names as well via the sort querystring.

rhymes commented:
@omenking sorry, I should have told you where the tests are: `spec/requests/dashboard_spec.rb`

omenking commented:
Hey @benhalpern, specs are on their way today

omenking commented:
As promised here are the specs.
You will notice that use `comments_count` is what I use to assert the order every spec. I had originally intended to validate against the data in specific to the sort. So if it were page_views I would extracted the views count from the element to match against since this would result in easier to understand specs when they fail. However, since created_at is not displayed and no date is displayed when article is a draft I would have to then add elements previously not there and I didn't want to get into design territory which feels like a new ticket.

Also pleas consider that content creators probably would mind having a sort based on published_at date however I don't see such a column, it could be that this is `updated_at` though I would need to inspect to confirm. 

I could have added this but since I am eager to open a new ticket to have a toggle between draft and published I figured the scope of that feature would make sense there for sorting. 

<img width="883" alt="Screen Shot 2019-04-09 at 2 54 57 PM" src="https://user-images.githubusercontent.com/7776/55828213-e6514680-5ad9-11e9-9790-48ac8c19e594.png">


```ruby
require "rails_helper"

RSpec.describe "Sorting Dashboard Articles", type: :system, js: true do
  let(:user) { create(:user) }
  let(:article1) { create(:article, user_id: user.id, created_at: 1.day.ago, positive_reactions_count: 5, page_views_count: 0, comments_count: 100) }
  let(:article2) { create(:article, user_id: user.id, created_at: 2.days.ago, positive_reactions_count: 1, page_views_count: 10, comments_count: 0) }
  let(:article3) { create(:article, user_id: user.id, created_at: 3.days.ago, positive_reactions_count: 0, page_views_count: 1000, comments_count: 50) }

  before do
    sign_in user
    article1
    article2
    article3
  end

  it "shows articles sorted by default in created_at DESC" do
    visit "/dashboard"
    values = page.all(".single-article .comments-count span.value").map { |e| e.text.to_i }
    value1 = article1.comments_count # 1 days ago
    value2 = article2.comments_count # 2 days ago
    value3 = article3.comments_count # 3 days ago
    expect(values).to eql([value1, value2, value3])
  end

  it "shows articles sorted by created_at ASC" do
    visit "/dashboard?sort=creation-asc"
    values = page.all(".single-article .comments-count span.value").map { |e| e.text.to_i }
    value3 = article3.comments_count # 3 days ago
    value2 = article2.comments_count # 2 days ago
    value1 = article1.comments_count # 1 days ago
    expect(values).to eql([value3, value2, value1])
  end

  it "shows articles sorted by comments_count DESC" do
    visit "/dashboard?sort=comments-desc"
    values = page.all(".single-article .comments-count span.value").map { |e| e.text.to_i }
    value1 = article1.comments_count # 100
    value3 = article3.comments_count # 50
    value2 = article2.comments_count # 0
    expect(values).to eql([value1, value3, value2])
  end

  it "shows articles sorted by page_views_count ASC" do
    visit "/dashboard?sort=views-asc"
    values = page.all(".single-article .comments-count span.value").map { |e| e.text.to_i }
    value1 = article1.comments_count # 0
    value2 = article2.comments_count # 10
    value3 = article3.comments_count # 100
    expect(values).to eql([value1, value2, value3])
  end

  it "shows articles sorted by positive_reactions_count ASC" do
    visit "/dashboard?sort=reactions-asc"
    values = page.all(".single-article .comments-count span.value").map { |e| e.text.to_i }
    value3 = article3.comments_count # 0
    value2 = article2.comments_count # 1
    value1 = article1.comments_count # 5
    expect(values).to eql([value3, value2, value1])
  end
end
```

omenking commented:
<img width="298" alt="Screen Shot 2019-04-09 at 3 22 44 PM" src="https://user-images.githubusercontent.com/7776/55828877-6330f000-5adb-11e9-9df6-e4613d5c2666.png">

Okay, hold on I must be drugs here because I missed some obvious sorting features. Let me quickly add them in.

omenking commented:
Okay, so I added in `published_at`.
I could have done specs for every scenario eg, ASC, DESC but thought you may appreciate me not making your tests slow with overly through tests.

I also refactor the sort options into ArticleHelpers as I was a bit lazy about this previously.

And I actually am on 💊drugs because of a  🐞viral infection so thank goodness for linters.

lightalloy commented on app/controllers/dashboards_controller.rb:
```diff
@@ -5,16 +5,45 @@ class DashboardsController < ApplicationController
-      @articles = @user.organization.articles.order("created_at DESC").decorate
+      @articles =
+        case params[:sort]
+        when "creation-asc" then @user.organization.articles.order("created_at ASC").decorate
```
An idea for refactoring:
```ruby
module Dashboard
  class ArticlesQuery

    ALLOWED_FIELDS = %w[published_at, ...] # list of allowed to sort fields
    # user is actually user or org, probably there's a better name
    def initialize(user, field, order)
      @user = user
      @order = 'desc' unless order == 'asc'
      @field = 'published_at' unless ALLOWED_FIELDS.include?(field)
    end

    def articles
      user.articles.order("#{field} #{order}")
    end
 
    private
    
    attr_reader :user, :order, :field
  end
end
```
Call in the controller like `Dashboard::ArticlesQuery.new(user, field, order).articles`
That requires thoughts on separating `field` and `order`.

rhymes replied:
Really like your refactoring, there's too much logic in the controller right now IMO

> Call in the controller like Dashboard::ArticlesQuery.new(user, field, order).articles. That requires thoughts on separating field and order.

That's why I initially proposed `creation-asc` because it can easily be split in `field` and `order`

lightalloy replied:
Thanks. Splitting `field` and `order` seems easy. Then there is just a need to map received field names to a the db field names.

rhymes replied:
oh, and a small fix on a thing that I just noticed: query objects are supposed to take in input the relation, so they can be composable with other external conditions or other query objects, so instead of:

```ruby
Dashboard::ArticlesQuery.new(user, field, order).articles
```

I'd change it to

```ruby
Dashboard::ArticlesQuery.new(user.articles, field, order).sort
```

the query is all about articles, not the user. It also makes them way easier to test because you can pass any "Article" relation instead of the user object 🙂

lightalloy replied:
Nice improvement :+1:

omenking replied:
Here are my various issues with the suggested refactor:

#### Naming
Query in the name implies searching/filtering but be we are doing ordering so so naming it should be sorting or ordering to better describe the functionality.

#### Clever vs Clear
My code is more verbose but it was intentional to be easy to be understood by juniors who I can tell going through the various tickets in this codebase struggle to complete the most basic of tasks which I suspect is due to this codebase deviating from Rails conventions.

#### Convention Over Configuration

There are ways to refactoring this and it would be done as a scope since scoping is already used heavily used in this project

```ruby
scope :sorting, lambda { |value|
  kind, dir = value.split('-')

  return %w{asc desc}.include?(dir)

  column =
  case kind
  when "creation"  then "created_at"
  when "views"     then "page_views_count"
  when "reactions" then "positive_reactions_count"
  when "comments"  then "comments_count"
  when "published" then "published_at"
  end

  order("? ?",column, dir)
}
```

Which would now result in

```ruby
@user.organization.articles.sorting(params[:sort]).decorate

```

I personally don't like the use of any model scoping for maintenance reasons but its the lesser of two abstraction evils.

#### Not Escaping

This here,
`user.articles.order("#{field} #{order}")`
Is not escaping and needs to be as such
`user.articles.order("? ?"),field,order`

#### Test All Conditions Explicity

This only checks in the case of one state ie. `asc and defaulting else` and needs to check both states explicitly.

```ruby
@order = 'desc' unless order == 'asc'
```


```ruby
dir =
case params[:order]
  when 'asc' then "ASC"
  when 'desc' then "DESC"
  else
    'asc'
end
```

another way could have been which is a bit less explicit than the previous example

```ruby
order ||= 'asc'
order = params[:order] if %w[asc desc].include?(param[:order])
```

My appologies for the harsh criticism.

abraham replied:
An option to explore too would be using [ransack](https://github.com/activerecord-hackery/ransack). I've used it on several projects and it works pretty well.

omenking replied:
I'm looked at ransack.

My thoughts are it adds another DSL for juniors to come in and worry about.
I think we need a strong use-case beside this one ticket for the addition of such a gem.
This codebase could really use a pruning of gems.

rhymes replied:
> #### Naming
> 
> Query in the name implies searching/filtering but be we are doing ordering so so naming it should be sorting or ordering to better describe the functionality.

I think @lightalloy's was more a suggestion about how to refactor the code, than about the name of the query object itself. 

> #### Clever vs Clear
> 
> My code is more verbose but it was intentional to be easy to be understood by juniors who I can tell going through the various tickets in this codebase struggle to complete the most basic of tasks which I suspect is due to this codebase deviating from Rails conventions.

I agree that code should be kept as simple as possible but that doesn't mean foregoing abstractions and good practices. Having the whole sorting logic inside the controller doesn't make it more readable, just makes the controller longer. You can just can keep your code and put it in a query object that's testable outside the controller for example, if you'd prefer not to be "clever" about using an array or a hash for dispatching.

> I personally don't like the use of any model scoping for maintenance reasons but its the lesser of two abstraction evils.

I don't like this option because you're putting an if/else inside a scope, which doesn't actually make it more readable, so unless you add a scope for each case I'm not sure how this is in improvement.

omenking replied:
`so unless you add a scope for each case I'm not sure how this is in improvement.`

Look here how I've refactored it in the current version. I don't add a scope for each `where` case. The scope eliminates the case in the controller.

```ruby
 def show
    fetch_and_authorize_user

    target = @user
    target = @user.organization if @user&.organization && @user&.org_admin && params[:which] == "organization"
    @articles = target.articles.sorting(params[:sort]).decorate
    # Updates analytics in background if appropriate:
    ArticleAnalyticsFetcher.new.delay.update_analytics(current_user.id) if @articles
  end
```

omenking commented:
I've refactored this to use  a scope. I am hoping this is my last refactor on this ticket.

```ruby
  scope :sorting, lambda { |value|
    value ||= "creation-desc"
    kind, dir = value.split("-")

    dir = "desc" unless %w[asc desc].include?(dir)

    column =
      case kind
      when "creation"  then :created_at
      when "views"     then :page_views_count
      when "reactions" then :positive_reactions_count
      when "comments"  then :comments_count
      when "published" then :published_at
      else
        :created_at
      end

    order(column => dir.to_sym)
  }
```

I did not use 1.9 ruby lambda syntax to be congruent with existing lambdas.

I removed the following/followers code because I think originally this was all coded into action and then later new endpoints were created and this was just legacy code kicking around. I did some Q&A and research and could not find their usage anywhere.

```ruby
  def show
    fetch_and_authorize_user

    target = @user
    target = @user.organization if @user&.organization && @user&.org_admin && params[:which] == "organization"
    @articles = target.articles.sorting(params[:sort]).decorate
    # Updates analytics in background if appropriate:
    ArticleAnalyticsFetcher.new.delay.update_analytics(current_user.id) if @articles
  end
```

omenking commented:
@benhalpern alls checks passing, specs are in, awaiting your approval

Turnerj commented on app/views/dashboards/_dashboard_article.html.erb:
```diff
@@ -28,7 +28,30 @@
+        <span class='page-views-count'>
+          <span class="value">
+            <% if article.page_views_count > 100 %>
+               <span class="value">article.page_views_count</span>
```
I think this is a bug - that will just print the text "article.page_views_count". I think you want `<%= article.page_Views_count %>`

omenking replied:
I was on DEV.to today and was like, I hope that's not my bug. 😅

Turnerj replied:
Hahaha! I was on the dashboard just before the change checking stuff. Refreshed the page and there was the ability to sort - a change I knew I wanted but didn't know it was actually already been worked on. After a bit of playing with it I realised about the view count missing and took a guess it was related.

Oh well, the site is still up at least! 😂
